File: test/test_record.py
import random
import sys
import gtar

def randomPath(term=.25):
    """Generates a random (but valid) path name for a property given a
    chance to terminate after any segment"""
    validNameChars = ([chr(c) for c in range(ord('a'), ord('z')+1)] +
                      [chr(c) for c in range(ord('A'), ord('A')+1)] +
                      [chr(c) for c in range(ord('0'), ord('9')+1)] + ['.', '/'])
    meaningfulBits = ['frames', 'vars', '.f32', '.f64', '.u8', '.u64', '.i32',
                      '.i64', '.ind', '.uni']
    randomBits = [''.join(random.sample(validNameChars, random.randint(1, 10))) for _ in range(20)]
    allBits = meaningfulBits + randomBits

    result = ''

    while random.random() > term:
        seg = ''.join(random.sample(allBits, random.randint(1, 5)))
        result += seg + '/'

    # get rid of trailing slash
    result = result[:-1]

    return (result if result else randomPath(term=term))

def test_pathIdentity(n=100):
    """Creating a Record object from a reasonable path and asking for
    its path should always be identity"""
    success = True
    for _ in range(n):
        path = randomPath()
        rec = gtar.Record(path)
        reconstructedPath = rec.getPath()

        relPath = path.lstrip('/')

        if relPath != reconstructedPath:
            print('Error reconstructing path from Record!'
                  ' "{}" != "{}" [{}, {}, {}]'.format(
                      relPath, reconstructedPath, rec.getGroup(), rec.getName(), rec.getIndex()))
            success = False

    assert success
    return success

def test_pathKnown():
    success = True

    (ind, uni) = (gtar.Resolution.Individual, gtar.Resolution.Uniform)
    txt = gtar.Resolution.Text
    (f32, u8, u32) = (gtar.Format.Float32, gtar.Format.UInt8, gtar.Format.UInt32)

    recs = {
        ('', 'position', '1000', f32, ind): gtar.Record('frames/1000/position.f32.ind'),
        ('rigid', 'body', '', u32, ind): gtar.Record('rigid/body.u32.ind'),
        ('rigid', 'com', 'ab cd', f32, ind): gtar.Record('rigid/frames/ab cd/com.f32.ind'),
        ('', 'N', '', u32, uni): gtar.Record('N.u32.uni'),
        ('', 'test.txt', '', u8, txt): gtar.Record('test.txt'),
        ('hoomd', 'types.json', '1024', u8, txt): gtar.Record('hoomd/frames/1024/types.json')
        }

    for key in recs:
        (group, name, index, fmt, res) = key
        rec = recs[key]

        match = True
        match = match and rec.getGroup() == group
        match = match and rec.getName() == name
        match = match and rec.getIndex() == index
        match = match and rec.getFormat() == fmt
        match = match and rec.getResolution() == res

        if not match:
            print('Error reconstructing known path {} (reconstructed path '
                  '"{}")!'.format(key, rec.getPath()))

        success = success and match

    assert success
    return success

def main():
    success = True

    for test in [test_pathIdentity, test_pathKnown]:
        try:
            success = test() and success
        except AssertionError:
            success = False

    sys.exit(not success)

if __name__ == '__main__': main()

File: test/test_Record.cpp

#include "Record.hpp"

#include <iostream>
#include <string>

using namespace gtar;
using namespace std;

// Get rid of any number of initial slashes ('/') in a string
string skipAbsolute(const string &str)
{
    size_t start(str.find_first_not_of('/'));
    return str.substr(start);
}

bool checkstr(const string &one, const string &two, bool shouldEqual, int line)
{
    bool failure(shouldEqual ^ (one == two));

    if(failure)
        cerr << "Error in line " << line << ": \"" << one << "\" != \"" << two << "\"" << endl;

    return failure;
}

bool check(const Record &rec, const string group, const string name,
           const string index, Behavior behavior, Format format,
           Resolution resolution, bool shouldEqual, int line)
{
    const Record ref(group, name, index, behavior, format, resolution);
    bool failure(shouldEqual ^ (ref == rec));

    if(failure)
        cerr << "Error in line " << line << endl;

    return failure;
}

int main()
{
    int result(0);

    {
        string str("frames/100/log.txt");
        Record rec(str);
        result |= check(rec, "", "log.txt", "100", Discrete, UInt8, Text, true, __LINE__);
        result |= checkstr(skipAbsolute(str), rec.getPath(), true, __LINE__);
    }

    {
        string str("frames/100/log.txt");
        Record rec(str);
        result |= check(rec, "", "lo.txt", "100", Discrete, UInt8, Text, false, __LINE__);
        result |= checkstr(skipAbsolute(str), rec.getPath(), true, __LINE__);
    }

    {
        string str("frames/100/position.f32.ind");
        Record rec(str);
        result |= check(rec, "", "position", "100", Discrete, Float32, Individual, true, __LINE__);
        result |= checkstr(skipAbsolute(str), rec.getPath(), true, __LINE__);
    }

    {
        string str("velocity.u64.uni");
        Record rec(str);
        result |= check(rec, "", "velocity", "", Constant, UInt64, Uniform, true, __LINE__);
        result |= checkstr(skipAbsolute(str), rec.getPath(), true, __LINE__);
    }

    {
        string str("/vars/stdout.txt/12");
        Record rec(str);
        result |= check(rec, "", "stdout.txt", "12", Continuous, UInt8, Text, true, __LINE__);
        result |= checkstr(skipAbsolute(str), rec.getPath(), true, __LINE__);
    }

    {
        string str("/rigid_body/frames/0/angular_momentum.f64.ind");
        Record rec(str);
        result |= check(rec, "rigid_body", "angular_momentum", "0", Discrete, Float64, Individual, true, __LINE__);
        result |= checkstr(skipAbsolute(str), rec.getPath(), true, __LINE__);
    }

    {
        string str("frame/2000/position.f32.ind");
        Record rec(str);
        result |= check(rec, "frame/2000", "position", "", Constant, Float32, Individual, true, __LINE__);
        result |= checkstr(skipAbsolute(str), rec.getPath(), true, __LINE__);
    }

    return result;
}
